# Project Data Analysis with Python : E-Commerce Public Dataset

- Name: Nathanael Steven Soetrisno

## Question Analysis
1. Which payment method is the most frequently used by customers?
2. Which product categories generate the highest and lowest revenue?
3. What are the top-selling products in each geographical region?
4. What is the average delivery time for orders, and how does it vary across different regions?
5. How many customers made repeat purchases in the last six months?

# Setup

In [2]:
pip install unidecode

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
from scipy import stats

# Data Wragling

## Gathering Data

In [7]:
df_customers = pd.read_csv('data/customers_dataset.csv')
df_customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [9]:
df_geolocation = pd.read_csv('data/geolocation_dataset.csv')
df_geolocation.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [29]:
df_order_items = pd.read_csv('data/order_items_dataset.csv')
df_order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [30]:
df_order_payments = pd.read_csv('data/order_payments_dataset.csv')
df_order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [11]:
df_order_reviews = pd.read_csv('data/order_reviews_dataset.csv')
df_order_reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [22]:
df_orders = pd.read_csv('data/orders_dataset.csv')
df_orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [23]:
df_product_category = pd.read_csv('data/product_category_name_translation.csv')
df_product_category.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [24]:
df_products = pd.read_csv('data/products_dataset.csv')
df_products.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [25]:
df_sellers = pd.read_csv('data/sellers_dataset.csv')
df_sellers.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

## Assessing Data

### Missing Value Check

In [32]:
print("Total missing value in customers dataset :", df_customers.isnull().sum(), '\n')
print("Total missing value in geolocation dataset :", df_geolocation.isnull().sum(), '\n')
print("Total missing value in order items dataset :", df_order_items.isnull().sum(), '\n')
print("Total missing value in order payments dataset :", df_order_payments.isnull().sum(), '\n')
print("Total missing value in order reviews dataset :", df_order_reviews.isnull().sum(), '\n')
print("Total missing value in orders dataset :", df_orders.isnull().sum(), '\n')
print("Total missing value in prduct category name dataset :", df_product_category.isnull().sum(), '\n')
print("Total missing value in products dataset :", df_products.isnull().sum(), '\n')
print("Total missing value in sellers dataset :", df_sellers.isnull().sum(), '\n')

Total missing value in customers dataset : customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64 

Total missing value in geolocation dataset : geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64 

Total missing value in order items dataset : order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64 

Total missing value in order payments dataset : order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64 

Total missing value in order reviews dataset : review_id                      0
order_id                       0
review_score              

Summary - Missing value check 
| Dataset                     | Columns with Missing Values                                       | Number of Missing Values                |
|-----------------------------|-------------------------------------------------------------------|-----------------------------------------|
| Customer Dataset             | None                                                             | 0                                       |
| Order Dataset                | order_approved_at, order_delivered_carrier_date, order_delivered_customer_date | 160, 1783, 2965                        |
| Product Dataset              | product_category_name, product_name_length, product_description_length, product_photos_qty, product_weight_g, product_length_cm, product_height_cm, product_width_cm | 610, 610, 610, 610, 2, 2, 2, 2          |
| Seller Dataset               | None                                                             | 0                                       |
| Product Category Dataset     | None                                                             | 0                                       |
| Review Dataset               | review_comment_title, review_comment_message                     | 87656, 58247                            |
| Order Payments Dataset       | None                                                             | 0                                       |
| Order Items Dataset          | None                                                             | 0                                       |
| Geolocation Dataset          | None                                                             | 0                                       |

### Duplicated Data Check

In [35]:
print("Total duplicated data in customers dataset :", df_customers.duplicated().sum())
print("Total duplicated data in geolocation dataset :", df_geolocation.duplicated().sum())
print("Total duplicated data in order items dataset :", df_order_items.duplicated().sum())
print("Total duplicated data in order payments dataset :", df_order_payments.duplicated().sum())
print("Total duplicated data in order reviews dataset :", df_order_reviews.duplicated().sum())
print("Total duplicated data in orders dataset :", df_orders.duplicated().sum())
print("Total duplicated data in prduct category name dataset :", df_product_category.duplicated().sum())
print("Total duplicated data in products dataset :", df_products.duplicated().sum())
print("Total duplicated data in sellers dataset :", df_sellers.duplicated().sum())

Total duplicated data in customers dataset : 0
Total duplicated data in geolocation dataset : 261831
Total duplicated data in order items dataset : 0
Total duplicated data in order payments dataset : 0
Total duplicated data in order reviews dataset : 0
Total duplicated data in orders dataset : 0
Total duplicated data in prduct category name dataset : 0
Total duplicated data in products dataset : 0
Total duplicated data in sellers dataset : 0


Summary - Duplicate Data Check

| Dataset                          | Number of Duplicated Records |
|-----------------------------------|------------------------------|
| Customers Dataset                 | 0                            |
| Geolocation Dataset               | 261,831                      |
| Order Items Dataset               | 0                            |
| Order Payments Dataset            | 0                            |
| Order Reviews Dataset             | 0                            |
| Orders Dataset                    | 0                            |
| Product Category Name Dataset     | 0                            |
| Products Dataset                  | 0                            |
| Sellers Dataset                   | 0                            |

### Outlier Check

Beberapa dataset tidak memiliki kolom numerik yang relevan untuk outliers
1. Customers dataset
2. Order reviews dataset
3. Sellers dataset

In [37]:
def outliers(data):
    q25, q75 = np.percentile(data,25), np.percentile(data,75)
    iqr = q75 - q25
    cut_off = iqr * 1.5
    minimum, maximum = q25 - cut_off, q75 + cut_off

    outliers = [x for x in data if x < minimum or x > maximum]
    return outliers

# Geolocation dataset - outliers pada latitude dan longitude
geolocation_lat_outliers = outliers(df_geolocation['geolocation_lat'])
geolocation_lng_outliers = outliers(df_geolocation['geolocation_lng'])
print(f"Total outliers in geolocation latitude: {len(geolocation_lat_outliers)}")
print(f"Total outliers in geolocation longitude: {len(geolocation_lng_outliers)}")

# Order items dataset - outliers pada price dan freight value
price_outliers = outliers(df_order_items['price'])
freight_value_outliers = outliers(df_order_items['freight_value'])
print(f"Total outliers in order item price: {len(price_outliers)}")
print(f"Total outliers in freight value: {len(freight_value_outliers)}")

# Order payments dataset - outliers pada payment_value
payment_value_outliers = outliers(df_order_payments['payment_value'])
print(f"Total outliers in payment value: {len(payment_value_outliers)}")

# Orders dataset - outliers pada delivery and estimated delivery times (days difference)
df_orders['delivery_time'] = (pd.to_datetime(df_orders['order_delivered_customer_date']) - pd.to_datetime(df_orders['order_purchase_timestamp'])).dt.days
delivery_time_outliers = outliers(df_orders['delivery_time'].dropna())
print(f"Total outliers in delivery time: {len(delivery_time_outliers)}")

# Products dataset - outliers pada weight, length, height, dan width
product_weight_outliers = outliers(df_products['product_weight_g'])
product_length_outliers = outliers(df_products['product_length_cm'])
product_height_outliers = outliers(df_products['product_height_cm'])
product_width_outliers = outliers(df_products['product_width_cm'])

print(f"Total outliers in product weight: {len(product_weight_outliers)}")
print(f"Total outliers in product length: {len(product_length_outliers)}")
print(f"Total outliers in product height: {len(product_height_outliers)}")
print(f"Total outliers in product width: {len(product_width_outliers)}")

Total outliers in geolocation latitude: 168240
Total outliers in geolocation longitude: 42348
Total outliers in order item price: 8427
Total outliers in freight value: 12134
Total outliers in payment value: 7981
Total outliers in delivery time: 5025
Total outliers in product weight: 0
Total outliers in product length: 0
Total outliers in product height: 0
Total outliers in product width: 0


Summary - Outlier Check

| Feature                     | Number of Outliers |
|-----------------------------|--------------------|
| Geolocation Latitude         | 168,240            |
| Geolocation Longitude        | 42,348             |
| Order Item Price             | 8,427              |
| Freight Value                | 12,134             |
| Payment Value                | 7,981              |
| Delivery Time                | 5,025              |
| Product Weight               | 0                  |
| Product Length               | 0                  |
| Product Height               | 0                  |
| Product Width                | 0                  |

### Cleaning Data

1. Orders dataset

In [45]:
df_orders['order_approved_at'].mode()

0    2018-02-27 04:31:10
Name: order_approved_at, dtype: object